File: bindings/nim/src/proven/safe_url.nim
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
#
# Safe URL parsing and validation.
# Thin wrapper over libproven FFI -- all logic lives in Idris.

import std/options
import lib_proven

type
  SafeUrl* = object
    ## A validated URL with parsed components.
    scheme*: string
    host*: string
    port*: Option[uint16]
    path*: string
    query*: string
    fragment*: string

proc parseUrl*(raw: string): Option[SafeUrl] =
  ## Parse a URL string into a SafeUrl.
  ## Returns None if the URL is invalid.
  if raw.len == 0:
    return none(SafeUrl)
  var res = provenUrlParse(unsafeAddr raw[0], csize_t(raw.len))
  if res.status != PROVEN_OK:
    return none(SafeUrl)

  # Marshal C strings to Nim strings before freeing
  let scheme = if res.components.scheme != nil and res.components.scheme_len > 0:
    $res.components.scheme
  else: ""
  let host = if res.components.host != nil and res.components.host_len > 0:
    $res.components.host
  else: ""
  let port = if res.components.has_port: some(res.components.port) else: none(uint16)
  let path = if res.components.path != nil and res.components.path_len > 0:
    $res.components.path
  else: ""
  let query = if res.components.query != nil and res.components.query_len > 0:
    $res.components.query
  else: ""
  let fragment = if res.components.fragment != nil and res.components.fragment_len > 0:
    $res.components.fragment
  else: ""

  provenUrlFree(addr res.components)

  some(SafeUrl(
    scheme: scheme,
    host: host,
    port: port,
    path: path,
    query: query,
    fragment: fragment
  ))

proc isValidUrl*(raw: string): bool =
  ## Check if a string is a valid URL.
  parseUrl(raw).isSome

proc isHttps*(url: SafeUrl): bool =
  ## Check if the URL uses HTTPS.
  url.scheme == "https"

proc isSecure*(url: SafeUrl): bool =
  ## Check if the URL uses a secure scheme.
  url.scheme in ["https", "ftps"]

proc hasPort*(url: SafeUrl): bool =
  ## Check if the URL has an explicit port.
  url.port.isSome

proc toString*(url: SafeUrl): string =
  ## Convert a SafeUrl back to a string.
  result = url.scheme & "://" & url.host
  if url.port.isSome:
    result.add(":" & $url.port.get())
  result.add(url.path)
  if url.query.len > 0:
    result.add("?" & url.query)
  if url.fragment.len > 0:
    result.add("#" & url.fragment)

File: bindings/csharp/src/Proven/SafeUrl.cs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

// SafeUrl.cs - URL parsing and validation.
//
// Thin P/Invoke wrapper over libproven. ALL computation is performed in
// verified Idris 2 code via the Zig FFI bridge. No logic is reimplemented here.

using System;
using System.Runtime.InteropServices;

namespace Proven
{
    /// <summary>
    /// Parsed URL components returned by <see cref="SafeUrl.Parse"/>.
    /// Matches the C ABI ProvenUrlComponents layout.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    internal struct NativeUrlComponents
    {
        public IntPtr Scheme;
        public nuint SchemeLen;
        public IntPtr Host;
        public nuint HostLen;
        public ushort Port;
        [MarshalAs(UnmanagedType.U1)]
        public bool HasPort;
        public IntPtr Path;
        public nuint PathLen;
        public IntPtr Query;
        public nuint QueryLen;
        public IntPtr Fragment;
        public nuint FragmentLen;
    }

    /// <summary>
    /// Native URL result struct matching C ABI ProvenUrlResult.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    internal struct NativeUrlResult
    {
        public int Status;
        public NativeUrlComponents Components;
    }

    /// <summary>
    /// Managed representation of parsed URL components.
    /// </summary>
    public sealed class ParsedUrl
    {
        /// <summary>URL scheme (e.g. "https").</summary>
        public string Scheme { get; }
        /// <summary>Host component.</summary>
        public string Host { get; }
        /// <summary>Port number, or null if not specified.</summary>
        public ushort? Port { get; }
        /// <summary>Path component.</summary>
        public string Path { get; }
        /// <summary>Query string, or null if not present.</summary>
        public string? Query { get; }
        /// <summary>Fragment identifier, or null if not present.</summary>
        public string? Fragment { get; }

        internal ParsedUrl(string scheme, string host, ushort? port,
                           string path, string? query, string? fragment)
        {
            Scheme = scheme;
            Host = host;
            Port = port;
            Path = path;
            Query = query;
            Fragment = fragment;
        }
    }

    /// <summary>
    /// Safe URL operations backed by formally verified Idris 2 code.
    /// Provides URL parsing and validation. All methods delegate to the
    /// libproven FFI.
    /// </summary>
    public static class SafeUrl
    {
        /// <summary>
        /// Parse a URL string into its components.
        /// Delegates to proven_url_parse via FFI. Frees native resources after
        /// copying data into managed memory.
        /// </summary>
        /// <param name="url">The URL string to parse.</param>
        /// <returns>Parsed URL components, or null on parse failure.</returns>
        public static ParsedUrl? Parse(string url)
        {
            byte[] bytes = MarshalHelpers.ToUtf8(url);
            IntPtr resultPtr = LibProven.proven_url_parse(bytes, (nuint)bytes.Length);

            if (resultPtr == IntPtr.Zero)
            {
                return null;
            }

            try
            {
                var result = Marshal.PtrToStructure<NativeUrlResult>(resultPtr);

                if (result.Status != 0)
                {
                    return null;
                }

                var components = result.Components;

                string scheme = components.Scheme != IntPtr.Zero
                    ? Marshal.PtrToStringUTF8(components.Scheme, (int)components.SchemeLen) ?? ""
                    : "";

                string host = components.Host != IntPtr.Zero
                    ? Marshal.PtrToStringUTF8(components.Host, (int)components.HostLen) ?? ""
                    : "";

                ushort? port = components.HasPort ? components.Port : null;

                string path = components.Path != IntPtr.Zero
                    ? Marshal.PtrToStringUTF8(components.Path, (int)components.PathLen) ?? ""
                    : "";

                string? query = components.Query != IntPtr.Zero && components.QueryLen > 0
                    ? Marshal.PtrToStringUTF8(components.Query, (int)components.QueryLen)
                    : null;

                string? fragment = components.Fragment != IntPtr.Zero && components.FragmentLen > 0
                    ? Marshal.PtrToStringUTF8(components.Fragment, (int)components.FragmentLen)
                    : null;

                return new ParsedUrl(scheme, host, port, path, query, fragment);
            }
            finally
            {
                LibProven.proven_url_free(resultPtr);
            }
        }

        /// <summary>
        /// Check if a URL string is valid (can be parsed).
        /// Delegates to proven_url_parse via FFI.
        /// </summary>
        /// <param name="url">The URL string to validate.</param>
        /// <returns>true if the URL is valid and can be parsed.</returns>
        public static bool IsValid(string url)
        {
            return Parse(url) is not null;
        }

        /// <summary>
        /// Extract the host component from a URL.
        /// Delegates to proven_url_parse via FFI.
        /// </summary>
        /// <param name="url">The URL string.</param>
        /// <returns>The host component, or null on parse failure.</returns>
        public static string? GetHost(string url)
        {
            return Parse(url)?.Host;
        }

        /// <summary>
        /// Extract the scheme component from a URL.
        /// Delegates to proven_url_parse via FFI.
        /// </summary>
        /// <param name="url">The URL string.</param>
        /// <returns>The scheme component, or null on parse failure.</returns>
        public static string? GetScheme(string url)
        {
            return Parse(url)?.Scheme;
        }

        /// <summary>
        /// Check if a URL uses the HTTPS scheme.
        /// Delegates to proven_url_parse via FFI.
        /// </summary>
        /// <param name="url">The URL string.</param>
        /// <returns>true if the scheme is "https".</returns>
        public static bool IsHttps(string url)
        {
            string? scheme = GetScheme(url);
            return string.Equals(scheme, "https", StringComparison.OrdinalIgnoreCase);
        }
    }
}
